pkg/util/instancemetadata/prod.go

```diff
@@ -79,7 +82,55 @@ func (p *prod) populateTenantIDFromMSI(ctx context.Context) error {
 	return nil
 }
 
+// populateInstanceMetadataFromEnvironment attempts to populate instance metadata from environment, it does not throw errors.
+func (p *prod) populateInstanceMetadataFromEnvironment() bool {
+	environment := azure.PublicCloud
+	if value, found := os.LookupEnv("AZURE_ENVIRONMENT"); found {
+		var err error
+		environment, err = azure.EnvironmentFromName(value)
+		if err == nil {
+			p.environment = &environment
+		}
+	}
+
+	// Getenv returns empty string if not set which indicates no value for the struct
+	p.subscriptionID = os.Getenv("AZURE_SUBSCRIPTION_ID")
+	p.tenantID = os.Getenv("AZURE_TENANT_ID")
+	p.location = os.Getenv("LOCATION")
+	p.resourceGroup = os.Getenv("RESOURCEGROUP")
+	p.hostname = os.Getenv("HOSTNAME_OVERRIDE")
+
+	if p.hostname == "" {
+		hostname, err := os.Hostname()
+		if err == nil {
+			p.hostname = hostname
+		}
+	}
+
+	return p.haveIntanceMetadata()
+}
+
+func (p *prod) haveIntanceMetadata() bool {
+	return p.environment != nil &&
+		p.subscriptionID != "" &&
+		p.tenantID != "" &&
+		p.location != "" &&
+		p.resourceGroup != "" &&
+		p.hostname != ""
+}
+
 func (p *prod) populateInstanceMetadata() error {
+	if os.Getenv("AZURE_EV2") != "" {
+		if p.populateInstanceMetadataFromEnvironment() {
+			// everything was populated from environment, can stop.
+			p.log.Info("created InstanceMetadata from Environment, not using http metadata endpoint")
+			return nil
+		}
+	}
+
+	// didn't find in env vars, build form metadata
+	// note this doesn't work in ACI, used with Ev2
```

Comment:
We currently have 2 implementations of `InstanceMetadata` interface. Can we have 3 instead?

* `pkg/util/instancemetadata/dev.go` - leave it as is.
* `pkg/util/instancemetadata/prod.go` - leave it mostly as is, but maybe rename into `publicProd` or something more appropriate.
* `pkg/util/instancemetadata/mynewprod.go` - create a new one which will just be reading from env. I guess it will be minimal: reading from `os.Getenv` + validating env name with `azure.EnvironmentFromName(value)`. We can make it testable by mocking `os.Getenv` in a similar we mock `time.Now` (example [code](https://github.com/Azure/ARO-RP/blob/924235001fd72793cc9770426dd74821c1e36bce/pkg/util/clientauthorizer/arm.go#L36) and [test](https://github.com/Azure/ARO-RP/blob/924235001fd72793cc9770426dd74821c1e36bce/pkg/util/clientauthorizer/arm_test.go#L124)).


And `pkg/util/instancemetadata/instancemetadata.go` will have something like:

```go
func New(ctx context.Context, isLocalDevelopmentMode bool) (InstanceMetadata, error) {
	if isLocalDevelopmentMode {
		return NewDev(true)
	}

	if isEv2 { // os.Getenv("AZURE_EV2") != "" or something similar? 
		return NewMyNewProd(true)
	}

	return newProd(ctx)
}
```

Reply:
This sounds like a better approach from one I did